### PR TITLE
bench: separately measure up/downstream time for netlink ipc

### PR DESCRIPTION
**Pull request by aditiharini**
Changes:
- modified the message the nltest module sends to portus to include the time when the message was received and sent back to portus.
- modified `src/bin/ipc_latency.rs` to allow the new message to be deserialized and use it to measure upstream and downstream times separately for the netlink IPC
- modified console output to reflect the new information and formatted so that it could be directly ingested by the plotting script: [https://github.com/ccp-project/eval-scripts/blob/master/plot/ipc_latency.r](url)
- modified the `Makefile` to use the `ipc_latency.rs` as the test for netlink IPC instead of `src/bin/nltest.rs`

### CODE REVIEW

File: src/bin/ipc_latency.rs [hunk 1]
```diff
@@ -9,9 +9,10 @@ extern crate time;
 use bytes::{ByteOrder, LittleEndian};
 use clap::Arg;
 use portus::ipc::{Backend, BackendSender, Blocking, Ipc, Nonblocking};
-use std::sync::{atomic, Arc};
 use time::Duration;
+use std::sync::{Arc, atomic};
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: keep this alphabetical? rust.vim should do this for you

File: src/bin/ipc_latency.rs [hunk 2]
```diff
@@ -43,30 +44,75 @@ impl portus::serialize::AsRawMsg for TimeMsg {
         Ok(TimeMsg(time::Timespec::new(sec, nsec)))
     }
 }
+#[derive(Debug)]
+struct DetTimeMsg(time::Timespec, time::Timespec);
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: What's "Det"?

File: src/bin/ipc_latency.rs [hunk 3]
```diff
@@ -43,30 +44,75 @@ impl portus::serialize::AsRawMsg for TimeMsg {
...
 use std::sync::mpsc;
-fn bench<T: Ipc>(b: &BackendSender<T>, mut l: Backend<T>, iter: u32) -> Vec<Duration> {
-    (0..iter)
+use portus::serialize::AsRawMsg;
+fn bench<T: Ipc>(
+    b: BackendSender<T>,
+    mut l: Backend<T>,
+    iter: u32,
+) -> Vec<Duration> {
+
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: extra newline

File: src/bin/ipc_latency.rs [hunk 4]
```diff
@@ -43,30 +44,75 @@ impl portus::serialize::AsRawMsg for TimeMsg {
...
+    let res = (0..iter)
         .map(|_| {
+
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: extra newline

File: src/bin/ipc_latency.rs [hunk 5]
```diff
@@ -43,30 +44,75 @@ impl portus::serialize::AsRawMsg for TimeMsg {
...
             }
         })
-        .collect()
+        .collect();
+    return res;
+
 }
 
+
+
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: more extra newlines

File: src/bin/ipc_latency.rs [hunk 6]
```diff
@@ -43,30 +44,75 @@ impl portus::serialize::AsRawMsg for TimeMsg {
...
 macro_rules! netlink_bench {
-    ($name: ident, $mode: ident) => {
+    ($name: ident, $mode: ident) => (
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: probably safer to use `({` in the macro definition?

File: src/bin/ipc_latency.rs [hunk 7]
```diff
@@ -43,30 +44,75 @@ impl portus::serialize::AsRawMsg for TimeMsg {
...
         #[cfg(target_os = "linux")] // netlink is linux-only
-        fn $name(iter: u32) -> Vec<Duration> {
+        fn $name(iter: u32) -> Vec<(Duration, Duration, Duration)> {
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: might be worth it to define a type for `(Duration, Duration, Duration)`

File: src/bin/ipc_latency.rs [hunk 8]
```diff
@@ -88,18 +134,45 @@ macro_rules! netlink_bench {
...
+            let (tx, rx) = mpsc::channel::<Vec<(Duration, Duration, Duration)>>();
 
             // listen
             let c1 = thread::spawn(move || {
                 let mut buf = [0u8; 1024];
-                let nl = portus::ipc::netlink::Socket::<$mode>::new()
-                    .map(|sk| {
-                        Backend::new(sk, Arc::new(atomic::AtomicBool::new(true)), &mut buf[..])
-                    })
+                // What does the backend represent?
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: `Backend` represents a wrapper around `Ipc` impls that deals in messages instead of bytes (i.e. a serialization wrapper layer).

On the receiving side, it also deals with message batching.

Comment (akshayknarayan) [reply src/bin/ipc_latency.rs]: But this comment should probably be removed

File: src/bin/ipc_latency.rs [hunk 9]
```diff
@@ -88,18 +134,45 @@ macro_rules! netlink_bench {
...
+                let mut nl = portus::ipc::netlink::Socket::<$mode>::new()
+                    .map(|sk| Backend::new(sk, Arc::new(atomic::AtomicBool::new(true)), &mut buf[..]))
                     .expect("nl ipc initialization");
                 tx.send(vec![]).expect("ok to insmod");
-                tx.send(bench(&nl.sender(), nl, iter)).expect("report rtts");
+                // this is what I added
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: remove this comment

File: src/bin/ipc_latency.rs [hunk 10]
```diff
@@ -88,18 +134,45 @@ macro_rules! netlink_bench {
...
+                nl.next().expect("receive echo");
+
+                let sender = nl.sender();
+                let res  = (0..iter)
+                .map(|_| {
+
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: \n

File: src/bin/ipc_latency.rs [hunk 11]
```diff
@@ -88,18 +134,45 @@ macro_rules! netlink_bench {
...
+                    let portus_send_time= time::get_time(); 
+                    let mut portus_recv_time = time::get_time();
+                    let mut kern_recv_time = time::get_time();
+                    let mut kern_send_time = time::get_time();
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: instead of initializing `mut` variables, have them be returned from the `if let` below - `mut` variables are harder to reason about and all the `time::get_time()` which are then overwritten is hard to parse.

Something like:
```rust
let (portus_recv_time, kern_recv_time, kern_send_time) = if let portus::serialize::Msg::Other(raw) = nl.next().expect("blah") {
  let portus_rt = get_time();
  let let kern_recv_msg = DetTimeMsg::from_raw_msg(raw).expect("get time from raw");
  (portus_rt, kern_recv_msg.0, kern_recv_msg.1)
} else {
  panic!(..)
};
```

File: src/bin/ipc_latency.rs [hunk 12]
```diff
@@ -88,18 +134,45 @@ macro_rules! netlink_bench {
...
+                    {
+                        portus_recv_time = time::get_time();
+                        let kern_recv_msg = DetTimeMsg::from_raw_msg(raw).expect("get time from raw");
+                        kern_recv_time = kern_recv_msg.0;
+                        kern_send_time = kern_recv_msg.1;
+
+                    } else {
+                        panic!("wrong type");
+                    }
+                    return (portus_recv_time-portus_send_time, kern_recv_time-portus_send_time, portus_recv_time-kern_send_time)
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: spaces between infix operators please, and should probably split this into 3 lines

File: src/bin/ipc_latency.rs (hunk 6 above)
Comment (akshayknarayan) [reply src/bin/ipc_latency.rs]: nope nevermind

File: src/bin/ipc_latency.rs [hunk 13]
```diff
@@ -243,61 +310,70 @@ fn main() {
         .version("0.1.0")
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: probably bump version

File: src/bin/ipc_latency.rs [hunk 14]
```diff
@@ -243,61 +310,70 @@ fn main() {
...
-        )
+        .arg(Arg::with_name("iterations")
+             .long("iterations")
+             .short("i")
+             .help("Specifies how many trials to run (default 100)")
+             .default_value("100"))
+        .arg(Arg::with_name("implementation")
+            .long("impl")
+            .help("Specifies the type of ipc being benchmarked")
+            .default_value("nl"))
```
Comment (akshayknarayan) [inline src/bin/ipc_latency.rs]: Instead of a string value here, it's probably best to use this: https://docs.rs/clap/2.32.0/clap/macro.arg_enum.html - that way the value of the string is automatically validated and we can match against it.

Also, should probably allow this arg to be given multiple times: https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.multiple

File: Makefile [hunk 15]
```diff
@@ -13,8 +13,9 @@ test-portus: build
 
 test-ipc: build
 ifeq ($(OS), Linux)
-	sudo ./target/debug/nltest
-	sudo ./target/debug/kptest
+	sudo ./target/debug/ipc_latency -i 10 --impl unix
+	sudo ./target/debug/ipc_latency -i 10 --impl nl
+	sudo ./target/debug/ipc_latency -i 10 --impl kp
```
Comment (akshayknarayan) [inline Makefile]: What is the story on nltest/kptest? Is it going away? (seems to be, since it's removed here)

This commit should probably remove those files and have some explanatory text in the commit message notes saying how this is tested now.

Comment (aditiharini) [reply Makefile]: I haven't made any changes to kptest, but I looked at kptest and it seems to do something very similar to what ipc_latency does when running the kp benchmark. I figured if we're using ipc_latency for nltest, we might as well use it for kptest too unless there's a good reason not to.

Comment (fcangialosi) [reply Makefile]: If you add functionality to test the character device (kp) in ipc_latency then it's fine to remove kptest, but otherwise I wouldn't remove it until then. We still need something to test whether or not the character device works.

File: src/bin/ipc_latency.rs (hunk 2 above)
Comment (aditiharini) [reply src/bin/ipc_latency.rs]: It's supposed to be short for "detailed" but I can see how that's a bad name.

File: src/bin/ipc_latency.rs (hunk 14 above)
Comment (aditiharini) [reply src/bin/ipc_latency.rs]: Sounds good.

File: src/bin/ipc_latency.rs (hunk 7 above)
Comment (aditiharini) [reply src/bin/ipc_latency.rs]: Yea I think that's a good idea so it's clear what each field represents.

Comment (aditiharini) [reply src/bin/ipc_latency.rs]: I may define a struct for the purpose I mentioned above.

File: src/bin/ipc_latency.rs (hunk 11 above)
Comment (aditiharini) [reply src/bin/ipc_latency.rs]: Yes. This makes a lot of sense. I'll change it.

File: Makefile (hunk 15 above)
Comment (aditiharini) [reply Makefile]: I'll move the kptest stuff to a different commit, since I nothing else I changed actually affects the kptest.

File: src/bin/ipc_latency.rs (hunk 7 above)
Comment (fcangialosi) [reply src/bin/ipc_latency.rs]: not strictly necessary, but might be better to name the fields just so it's clear, just wrapping it in a type doesn't really add any information about what the individual fields mean
